### ISSUE TITLE
DiskReadViolation triggered by setCrashlyticsCollectionEnabled() 

### ISSUE THREAD

**Issue by DanielNovak** [firebase/firebase-android-sdk#3265]
Describe your environment

  * Android Studio version: 2020.3.1 Patch 4
  * Firebase Component: Crashlytics
  * Component version: 18.2.6

Describe the problem

Firebase Crashlytics is triggering Android StrictMode DiskReadViolation when `setCrashlyticsCollectionEnabled(true)` is called.

```
StrictMode policy violation; ~duration=351 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1659)
        at libcore.io.BlockGuardOs.access(BlockGuardOs.java:74)
        at libcore.io.ForwardingOs.access(ForwardingOs.java:131)
        at android.app.ActivityThread$AndroidOs.access(ActivityThread.java:7715)
        at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:281)
        at java.io.File.exists(File.java:813)
        at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:738)
        at android.app.SharedPreferencesImpl.access$900(SharedPreferencesImpl.java:59)
        at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:672)
        at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:691)
        at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:59)
        at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:604)
        at com.google.firebase.crashlytics.internal.common.DataCollectionArbiter.storeDataCollectionValueInSharedPreferences(DataCollectionArbiter.java:206)
        at com.google.firebase.crashlytics.internal.common.DataCollectionArbiter.setCrashlyticsDataCollectionEnabled(DataCollectionArbiter.java:92)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsCore.setCrashlyticsCollectionEnabled(CrashlyticsCore.java:257)
        at com.google.firebase.crashlytics.FirebaseCrashlytics.setCrashlyticsCollectionEnabled(FirebaseCrashlytics.java:448)
```

#### Steps to reproduce:

Enable strict mode:
```
        StrictMode.setThreadPolicy(
            ThreadPolicy.Builder()
                .detectAll()
                .penaltyLog()
                .build()
```
and call:

`FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(enableCrashlytics)`

#### Relevant Code:

The problem is in Firebase `DataCollectionArbitrer.java`. It's using `commit()` instead of `apply()`:

```
    private static void storeDataCollectionValueInSharedPreferences(SharedPreferences sharedPreferences, Boolean enabled) {
        Editor prefsEditor = sharedPreferences.edit();
        if (enabled != null) {
            prefsEditor.putBoolean("firebase_crashlytics_collection_enabled", enabled);
        } else {
            prefsEditor.remove("firebase_crashlytics_collection_enabled");
        }

        prefsEditor.commit();
    }
```
For some reason the developers also marked it with `@SuppressLint({"ApplySharedPref"})`. So maybe they had a reason to do it like this? But then this call is always blocking the main thread on many apps on start...

**Comment by argzdev** [comment]
Hi @DanielNovak, thanks for reporting. We'll investigate this and see what we can find.

**Comment by argzdev** [comment]
@DanielNovak, I was unable to reproduce the issue with the given steps.

After enabling StrictMode policy. I've added the `FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(enableCrashlytics)` with either true or false.

However, I was not able to experience any issue on runtime on either Android 10.0, 11.0, or 12.0

Am I missing anything? If it's possible could you provide us a minimal reproducible example with the issue so I can investigate this further.

**Comment by DanielNovak** [comment]
@argzdev There is no direct issue or crash - but you will see the log entry in the logs. You can also change the code to:
```
StrictMode.setThreadPolicy(
            ThreadPolicy.Builder()
                .detectAll()
                .penaltyDeath()
                .build()
```
to crash the app on strict mode violation. But make sure that you enable StrictMode before calling `setCrashlyticsCollectionEnabled`. 

So this strict mode is more like a warning - that the Crashlytics is accessing the disk on main thread, but it's not causing any visible problems or crashes.

**Comment by argzdev** [comment]
Thanks for the clarification @DanielNovak, I see it now. I'll notify an engineer and see what we can do about this.

**Comment by mrichards** [comment]
Thanks @DanielNovak, I merged the change and it will be available in our next release, `18.2.7`, coming out in a couple weeks. I'll leave this ticket open until it ships.

**Comment by argzdev** [comment]
Since this has been released in our [latest version 18.2.7](https://firebase.google.com/support/release-notes/android#crashlytics_v18-2-7), I'll be closing this now. Thanks!
